plantdescriptionengine/src/main/java/eu/arrowhead/core/plantdescriptionengine/providedservices/pde_mgmt/PlantDescriptionValidator.java

```diff
@@ -0,0 +1,446 @@
+     * metadata combination.
+     */
+    private String uniqueIdentifier(final PdeSystem system) {
+        final String systemName = system.systemName().orElse("");
+        final Map<String, String> metadata = system.metadata();
+        String result = systemName + "{";
+        if (!metadata.isEmpty()) {
+            result += Metadata.toString(metadata);
+        }
+        return result + "}";
+    }
+
+    /**
+     * Ensure that each system in every given entry is uniquely identifiable,
+     * either by a name or by metadata.
+     *
+     * @param entries A set of Plant Description Entries.
+     */
+    private void ensureIdentifiableSystems(final Set<PlantDescriptionEntry> entries) {
+
+        final ArrayList<PdeSystem> systems = new ArrayList<>();
+
+        for (final PlantDescriptionEntry entry : entries) {
+            systems.addAll(entry.systems());
+        }
+
+        final Set<String> uniqueIdentifiers = new HashSet<>();
+
+        for (final PdeSystem system : systems) {
+            final String uid = uniqueIdentifier(system);
+            if (!uniqueIdentifiers.add(uid)) {
+                errors.add("System with ID '" + system.systemId() +
+                    "' cannot be uniquely identified by its name/metadata combination.");
+            }
+        }
+    }
+
+    /**
+     * Retrieves the complete include chain of the specified entry from among
+     * the given list of all entries.
+     *
+     * @param entry      Plant Description entry whose include chain will be
+     *                   retrieved.
+     * @param allEntries A list containing all Plant Description entries in
+     *                   which to search for included entries.
+     * @throws ValidationException If there is an include cycle, or if an
+     *                             included entry is not present in
+     *                             {@code allEntries}.
+     */
+    private Set<PlantDescriptionEntry> getIncludeChain(
+        final PlantDescriptionEntry entry,
+        final Map<Integer, ? extends PlantDescriptionEntry> allEntries
+    ) throws ValidationException {
+
+        final HashSet<PlantDescriptionEntry> visitedEntries = new HashSet<>();
+        final LinkedList<PlantDescriptionEntry> queue = new LinkedList<>();
+
+        queue.add(entry);
+
+        while (!queue.isEmpty()) {
+            final PlantDescriptionEntry nextEntry = queue.pop();
+
+            if (!visitedEntries.add(nextEntry)) {
+                throw new ValidationException("Error in include list: Cycle detected.");
+            }
+
+            for (final Integer includedId : nextEntry.include()) {
+                if (!allEntries.containsKey(includedId)) {
+                    throw new ValidationException("Error in include list: Entry '" + includedId + "' is required by entry '" + nextEntry.id() + "'.");
+                }
+                final var includedEntry = allEntries.get(includedId);
```

Comment:
My opinion that using var worsen code readability

Reply:
Fixed